src/lib/Inference.hs

```diff
@@ -1290,67 +1311,78 @@ checkULam (ULamExpr _ (UPatAnn p ann) body) piTy@(PiType (PiBinder _ _ arr) _ _)
         checkSigma body resultTy
 
 checkInstanceArgs
-  :: (EmitsBoth o, Inferer m)
+  :: (EmitsInf o, Inferer m, SinkableE e, SubstE Name e, HoistableE e)
   => Nest UPatAnnArrow i i'
-  -> (forall o'. (EmitsBoth o', Ext o o') =>  m i' o' (Atom o'))
-  -> m i o (Atom o)
-checkInstanceArgs Empty cont = cont
+  -> (forall o'. (EmitsInf o', Ext o o') =>  m i' o' (e o'))
+  -> m i o (Abs (Nest PiBinder) e o)
+checkInstanceArgs Empty cont = Abs Empty <$> cont
 checkInstanceArgs (Nest (UPatAnnArrow (UPatAnn p ann) arrow) rest) cont = do
   case arrow of
     ImplicitArrow -> return ()
     ClassArrow    -> return ()
+    PlainArrow    -> return ()
     _ -> throw TypeErr $ "Not a valid arrow for an instance: " ++ pprint arrow
-  checkAnnWithMissingDicts ann \ds getArgTy -> do
+  ab <- checkAnnWithMissingDicts ann \ds getArgTy -> do
     introDicts (eSetToList ds) $ do
       argTy <- getArgTy
-      buildLamInf (getNameHint p) arrow argTy (const $ return Pure) \v -> do
-        bindLamPat p v $
+      buildPiAbsInf (getNameHint p) arrow argTy \v -> do
+        WithSrcB pos (UPatBinder b) <- return p  -- TODO: enforce this syntactically
```

Comment:
> Why? I don't see a reason why it has to be a plain binder. Can't we use bindLamPat like we used to?

The reason it has to be a plain binder is that other patterns would produce decls, and instances don't emit decls anymore.

> Oh wait. Why are we building up a nest of Pis and not lambdas?

The binders we're creating here are eventually going to end up in an `InstanceDef`, which we didn't have before. I used `PiBinder`s instead of `LamBinder`s for the arg binders in `InstanceDef` because they don't have effects, but it's true that they're more like lambda binders in spirit.

Reply:
> other patterns would produce decls

That's not necessarily true, is it? The main reason why `ProjectElt` exists is to allow us to refer to pieces of patterns without emitting decls.

Reply:
Right, but some patterns produce decls, like table patterns.